### PR TITLE
Add ldvirtftn

### PR DESCRIPTION
**Pull request by BadRyuner**
- Adds support for Ldvirtftn OpCode
- Adds test for Ldvirtftn OpCode using delegates

Also unexpected, but ((object)anyInteger).ToString() throws exception +_+

### CODE REVIEW

File: src/Platforms/Echo.Platforms.AsmResolver/Emulation/Dispatch/ObjectModel/LdvirtftnHandler.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            {
+                var functionPointer = methods.GetAddress(resolvedVirtualFunction);
+                stack.Push(factory.CreateNativeInteger(functionPointer), type);
+                return CilDispatchResult.Success();
+            }
+
+            // else switch to BaseType and try resolve again
+            thisObjectType = thisObjectType.BaseType?.Resolve();
+        }   // or exit and throw CilEmulationException
+        while (thisObjectType != null);
```
Comment (Washi1337) [inline src/Platforms/Echo.Platforms.AsmResolver/Emulation/Dispatch/ObjectModel/LdvirtftnHandler.cs:52]: This is not sufficient. Ldvirtftn follows the exact same method devirtualization process as `callvirt`, with all its edge cases.

**Review by Washi1337 — CHANGES_REQUESTED** [review]
This does not cover the semantics of `ldvirtftn` adequately. It has to follow the exact same behavior as `callvirt` when it comes to method devirtualization.

We may have to somehow extract the method devirtualization of `callvirt` into its own separate unit such that it can be reused by both `callvirt` and `ldvirtftn` without having to copy implementations.

**Comment by Washi1337** [comment]
> Also unexpected, but ((object)anyInteger).ToString() throws exception

If that is the case, we need a test case for this too.

**Comment by BadRyuner** [comment]
> > Also unexpected, but ((object)anyInteger).ToString() throws exception
> 
> If that is the case, we need a test case for this too.

```
System.ArgumentException: "The type parameter !0 could not be resolved to a concrete type argument in the current context."
```

```
- AsmResolver.DotNet.Memory.TypeAlignmentDetector.VisitGenericParameter(GenericParameterSignature signature) in AsmResolver.DotNet.Memory\TypeAlignmentDetector.cs:line 94
   - AsmResolver.DotNet.Signatures.Types.GenericParameterSignature.AcceptVisitor[TResult](ITypeSignatureVisitor`1 visitor) in AsmResolver.DotNet.Signatures.Types\GenericParameterSignature.cs:line 70
   - AsmResolver.DotNet.Memory.TypeAlignmentDetector.VisitTypeDefinition(TypeDefinition type) in AsmResolver.DotNet.Memory\TypeAlignmentDetector.cs:line 165
   - AsmResolver.DotNet.Memory.TypeMemoryLayoutDetector.VisitValueTypeDefinition(TypeDefinition type) in AsmResolver.DotNet.Memory\TypeMemoryLayoutDetector.cs:line 182
   - AsmResolver.DotNet.Memory.TypeMemoryLayoutDetector.VisitTypeDefinition(TypeDefinition type) in AsmResolver.DotNet.Memory\TypeMemoryLayoutDetector.cs:line 172
   - AsmResolver.DotNet.Memory.TypeMemoryLayoutDetector.VisitTypeDefOrRef(ITypeDefOrRef type) in AsmResolver.DotNet.Memory\TypeMemoryLayoutDetector.cs:line 147
   - AsmResolver.DotNet.Memory.TypeMemoryLayoutDetection.GetImpliedMemoryLayout(ITypeDefOrRef type, Boolean is32Bit) in AsmResolver.DotNet.Memory\TypeMemoryLayoutDetection.cs:line 29
   - Echo.Platforms.AsmResolver.Emulation.ValueFactory.GetTypeDefOrRefContentsLayout(ITypeDefOrRef type, GenericContext context, UInt32 alignment) in D:\Work\Echo\src\Platforms\Echo.Platforms.AsmResolver\Emulation\ValueFactory.cs:line 544
   - Echo.Platforms.AsmResolver.Emulation.ValueFactory.GetTypeSignatureContentsLayout(TypeSignature type) in D:\Work\Echo\src\Platforms\Echo.Platforms.AsmResolver\Emulation\ValueFactory.cs:line 603
   - Echo.Platforms.AsmResolver.Emulation.ValueFactory.GetTypeContentsMemoryLayout(ITypeDescriptor type) in D:\Work\Echo\src\Platforms\Echo.Platforms.AsmResolver\Emulation\ValueFactory.cs:line 498
   - Echo.Platforms.AsmResolver.Emulation.Dispatch.ObjectModel.NewObjHandler.HandleValueTypeNewObj(CilExecutionContext context, CilInstruction instruction, IMethodDescriptor constructor, TypeSignature instanceType, IList`1 arguments) in D:\Work\Echo\src\Platforms\Echo.Platforms.AsmResolver\Emulation\Dispatch\ObjectModel\NewObjHandler.cs:line 50
   - Echo.Platforms.AsmResolver.Emulation.Dispatch.ObjectModel.NewObjHandler.Dispatch(CilExecutionContext context, CilInstruction instruction) in D:\Work\Echo\src\Platforms\Echo.Platforms.AsmResolver\Emulation\Dispatch\ObjectModel\NewObjHandler.cs:line 27
   - Echo.Platforms.AsmResolver.Emulation.Dispatch.CilDispatcher.Dispatch(CilExecutionContext context, CilInstruction instruction) in D:\Work\Echo\src\Platforms\Echo.Platforms.AsmResolver\Emulation\Dispatch\CilDispatcher.cs:line 72
```

looks like a new issue for AsmResolver

**Comment by Washi1337** [comment]
> looks like a new issue for AsmResolver

If that is the case, it would be great if you could make a minimal reproducible test case and put it on the [AsmResolver issue board](https://github.com/Washi1337/AsmResolver/issues). 

However I often found this to be the case of Echo simply not always instantiating the generic types properly everywhere. In that case, we still need a minimal unit test, but specifically using Echo instead :)

**Comment by Washi1337** [comment]
This is very likely an Echo bug. I don't see any evidence of the actual type memory layout algorithm being problematic here.

This problem was partially addressed for the `call` opcode. See CallHandlerBase.cs below:

https://github.com/Washi1337/Echo/blob/687c1fed66e64b645804b6cc00537ede5f2f5bb6/src/Platforms/Echo.Platforms.AsmResolver/Emulation/Dispatch/ObjectModel/CallHandlerBase.cs#L151-L206

We should probably copy/extend this to work for fields as well.

Regardless, this is definitely out of scope for this PR, and really should be in a separate issue.

**Pull request closed without merging**